pyam/iiasa.py

```diff
@@ -37,31 +47,65 @@ def valid_connection_names():
 class Connection(object):
     """A class to facilitate querying an IIASA scenario explorer database"""
 
-    def __init__(self, name):
+    def __init__(self, name, creds=None, application=None):
         """
         Parameters
         ----------
         name : str
             A valid database name. For available options, see
             valid_connection_names().
+        creds : list-like or dict, optional
+            An ordered container with entries of 'username' and 'password',
+            or a dictionary with the same keys.
+        application : str, optional
+            If using non-valid connection name, the application name must
+            also be supplied.
         """
         valid = valid_connection_names()
-        if name not in valid:
-            raise ValueError('{} is not a valid name. Choose one of {}'.format(
-                name, valid))
-
-        logger().info(
-            'You are connected to the {} {}. Please cite as:\n\n{}'
-            .format(name, 'scenario explorer', _CITATIONS[name])
-        )
+        if name not in valid and application is None:
+            msg = '{} not recognized as a valid connection name. '
+            'Either *also* provide an "application" argument, '
+            'or choose from one of the supported connections: {}.'
+            raise ValueError(msg.format(name, valid))
+
+        if name in valid:
+            logger().info(
+                'You are connected to the {} {}. Please cite as:\n\n{}'
+                .format(name, 'scenario explorer', _CITATIONS[name])
+            )
 
         self.base_url = _URL_TEMPLATE.format(name)
 
-    @lru_cache()
+        # get authorization
+        if creds is None:  # anonymously
+            url = _LOGIN_URL.replace('login', 'anonym')
+            self._auth = requests.get(url).json()
+        else:  # read credentials from kwargs
+            try:
+                if isinstance(creds, collections.Mapping):
+                    user, pw = creds['username'], creds['password']
+                else:
+                    user, pw = creds
+            except Exception as e:
+                msg = 'Could not read credentials: {}\n{}'.format(
+                    creds, str(e))
+                raise type(e)(msg)
+            # request authentication
+            headers = {'Accept': 'application/json',
+                       'Content-Type': 'application/json'}
+            app = application or _LOGIN_APP_NAMES[name]
+            data = {'username': user, 'password': pw, 'application': app}
+            response = requests.post(
+                _LOGIN_URL, headers=headers, data=json.dumps(data)
+            )
+            self._auth = response.json()
+            if not response.ok:  # auth failed, provide user a nice message
+                msg = 'Login to {} failed with given credentials: {}'
```

Comment:
I wouldn't output the credentials, at least not the password

Reply:
Ah here, the creds aren't printed, only the app name and error log.

Reply:
Where `self._auth` is the result of a failed `requests.post()`